### DISCUSSION TITLE
Remote Server Administration Requirements

### DISCUSSION THREAD

**Discussion by Rob-NY** [jamulussoftware/jamulus#1863] [Ideas]
With a few proposals out there with respect to remote management and control of Jamulus servers, I thought I would start a solution-agnostic thread to discuss the requirements and use-cases associated with this need.

From my perspective, the remote management of a Jamulus server is not entirely about controlling Jamulus itself - but also about the environment and file-system on which it runs.  With this in mind, here is a quick list of functional elements that I've been considering from a management and operations perspective:

- Start / Re-start Jamulus server  (1)
- Get recording state  (2)
- Toggle recording state (2)
- Get clients connected w/ IP info (2)
- Block an IP address (1)
- Block all IP addresses except ... (1)
- Package/move recording directory contents elsewhere for delivery (3)
- Clear recording directory (3)
- Report free disk space (3)
- Server load/memory (1)
- Dynamically change welcome message (3)

In the list above, those with (1) are OS or platform dependent, (2) are directly related to Jamulus Server, and (3) are file-system operations.

Only a few of these are foundational to Jamulus itself.  Most are outside of Jamulus, but are still part of the overall management and administration needs.

My point is that remote admin goes beyond the Jamulus code.  Many items in the list above can't be (or shouldn't be) built into Jamulus -- nor should the remote administration requirements distract from the core Jamulus functionality and optimizations.  

With this list in mind (and as it changes), I suspect that a big chunk of remote administration will have nothing to do with the Jamulus code-base at all.  There clearly will need to be a different agent that serves this remote administration need and this agent will talk to the jamulus server as well as the OS, the cloud, a web UI, etc.   

Therefore, I think the first item to address is to determine what should be included in Jamulus in order to _support_ remote administration; and then how best to implement that without distracting from Jamulus's core focus.

**Comment by ann0see** [comment]
Yes, that's an interesting point (especially "Block an IP address"). This feature is something which was widely discussed (and I am on the pro side); however it is not part of the official code base yet (and is a big controversy in the community). You could get the IP from jamulus and "ban" it via firewall. 

Server load/memory should be monitored on OS level, not part of Jamulus; same for disk space.

As far as I see, your ideas are also somewhat connected to @gene96817 's proposal with jam management?

**Reply by Rob-NY** [reply]
Yes, all these discussion threads are inter-related.  But only a portion of their proposals should be part of core Jamulus.  As you mention, the IP blocking should be done outside of Jamulus as well as all the OS monitoring, file manipulation, etc.  So what needs to be decided is how **_Jamulus_** needs to be modified to support external management/operation systems.  This should be a very small and non-intrusive set of asks.

**Reply by gene96817** [reply]
These discussions open new options for how session management could be implemented. I agree with the goal of minimizing the impact on the core Jamulus code. 

It would be helpful to keep a clear separation between remote Jamulus administration vs remote server administration. Consider the case where there are multiple Jamulus servers on a single server or remotely managing a Jamulus server on a cloud service.
